engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnCompensationSubscriptionBehaviour.java

```diff
@@ -69,26 +73,28 @@ public void createCompensationSubscription(
     }
   }
 
-  public boolean triggerCompensation(final BpmnElementContext context) {
-
+  public boolean triggerCompensation(
+      final ExecutableFlowNode element, final BpmnElementContext context) {
     final Set<CompensationSubscription> completedCompensationSubscription =
         getCompensationSubscriptionForCompletedActivities(context);
 
     if (completedCompensationSubscription.isEmpty()) {
       return false;
     }
-    // activate the compensation handler
-    completedCompensationSubscription.forEach(
-        compensation -> {
-          final var key = compensation.getKey();
-          final var compensationRecord = compensation.getRecord();
-          compensationRecord
-              .setThrowEventId(BufferUtil.bufferAsString(context.getElementId()))
-              .setThrowEventInstanceKey(context.getElementInstanceKey());
-          stateWriter.appendFollowUpEvent(
-              key, CompensationSubscriptionIntent.TRIGGERED, compensationRecord);
 
-          activateCompensationHandler(compensationRecord.getCompensableActivityId(), context);
+    // trigger compensation on process level
+    triggerCompensationForScopeId(
+        context, BufferUtil.bufferAsString(element.getFlowScope().getId()));
+
+    final var completedSubprocessScopeId =
+        getCompletedSubprocessScopeId(context.getTenantId(), context.getProcessInstanceKey());
```

Comment:
Your suggestion is not clear to me, why we are filtering twice ? And why we need to write `triggered` event two times ?

Reply:
:sweat_smile: I had the idea to access the state less often and to remove the subprocess property from the record. This could be a minor improvement. 

Feel free to check my changes here if you like it: https://github.com/camunda/zeebe/commit/25554974fa931df124ead2015fc127393d24cf08